### PR TITLE
Patch with CUDA 11.2+ fix

### PR DESCRIPTION
**Pull request by jakirkham**
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

File: recipe/meta.yaml [hunk 1]
```diff
@@ -19,9 +19,11 @@ package:
 source:
   - url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
     sha256: {{ sha256 }}
+    patches:
+      - PR_5993.patch
```
Comment (jakirkham) [inline recipe/meta.yaml:23]: xref: https://github.com/cupy/cupy/pull/5993

**Comment by jakirkham** [comment]
@leofang please let me know if I missed anything here 🙂

**Comment by jakirkham** [comment]
Looks like there are conflicts applying the patch 😞

**Comment by jakirkham** [comment]
AFAICT this was mostly related to CI changes that we don't need here. So grabbed the first commit and tweaked it to include the flake8 fix. Though please let me know if anything important was missed

**Comment by jakirkham** [comment]
Patch applies! 😄

**Review by leofang — APPROVED** [review]
Thanks, John!

**Comment by jakirkham** [comment]
Thanks Leo! 😀

**Comment by leofang** [comment]
@jakirkham What happens if we don't relax the runtime `cudatoolkit` requirement here? I guess the applied repo-patch would not be immediately applicable to new packages, until someone reruns the repo-patch?

**Comment by jakirkham** [comment]
Good question. We should already be ok as we changed how the pinning is done for 11.2 in PR ( https://github.com/conda-forge/nvcc-feedstock/pull/71 )

**Comment by leofang** [comment]
Ah OK, I forgot about it 😅 Thanks!

**Comment by jakirkham** [comment]
No worries. There are a lot of moving pieces. Also it is helpful to revisit these things periodically 😄

**Comment by jakirkham** [comment]
Hmm...seeing lots of lines like this. Particularly in the ARM CUDA 11.2 builds

```
  /home/conda/feedstock_root/build_artifacts/cupy_1635988585483/work/install/../cupy/_core/include/cupy/complex/complex.h(94): warning: __host__ annotation is ignored on a function("complex") that is explicitly defaulted on its first declaration

  /home/conda/feedstock_root/build_artifacts/cupy_1635988585483/work/install/../cupy/_core/include/cupy/complex/complex.h(94): warning: __device__ annotation is ignored on a function("complex") that is explicitly defaulted on its first declaration
```

**Comment by leofang** [comment]
Yes this is known.. and partly my bad 😅 tl;dr is it is safe to ignore.

**Comment by jakirkham** [comment]
No worries. Think the issue is those builds are getting stuck in a loop they can't resolve. Can't tell if the warnings are the cause or merely a symptom (like some `include` that isn't being resolved well). In either case it seems like they are unlikely to complete. They've run for over 5hrs at this point and are very close to the CI time limit

**Comment by leofang** [comment]
If they hit timeout, usually the next run would success. It's awfully slow mainly because 1. they're in the emulator 2. too many template kernels x supported archs.

**Comment by jakirkham** [comment]
Actually it might not be an infinite loop, but we do seem to see more warnings on CUDA 11.2 builds than other CUDA versions. For example [this build]( https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=402746&view=logs&j=1e116cca-0a74-5ed2-d113-bc13ce542b6b&t=b9730c68-c606-502c-e481-92bd41fbcce7 ) demonstrates this issue.

Yeah agree the emulation slows things down quite a bit. Unfortunately we lack another option atm. Didn't know about the templating though that makes sense. Is the templating tied to CUDA version somehow?

Anyways I guess my inclination at this point would be to merge so we can at least get some of the packages out. We can restart CI on `master` to get any stragglers out as well.

**Comment by jakirkham** [comment]
Looks like almost all of them except one completed on `master`. Restarting the one failing job

**Comment by jakirkham** [comment]
Some folks internally asked we mark build number `0` broken. So submitted PR ( https://github.com/conda-forge/admin-requests/pull/338 ) to do that

**Comment by leofang** [comment]
Yup I'd have done it myself if I had time this morning, thanks John!

**Comment by jakirkham** [comment]
Glad to hear it. Just wanted to be transparent and offer people the chance to raise any concerns 🙂

**Pull request merged**